src/nemos/base_regressor.py

```diff
@@ -0,0 +1,465 @@
+        Default is `None`. If no solver specified, one will be chosen based on the regularizer.
+        Please see table above for regularizer/optimizer pairings.
+    solver_kwargs :
+        Optional dictionary for keyword arguments that are passed to the solver when instantiated.
+        E.g. stepsize, acceleration, value_and_grad, etc.
+         See the jaxopt documentation for details on each solver's kwargs: https://jaxopt.github.io/stable/
+
+    See Also
+    --------
+    Concrete models:
+
+    - [`GLM`](../glm/#nemos.glm.GLM): A feed-forward GLM implementation.
+    - [`GLMRecurrent`](../glm/#nemos.glm.GLMRecurrent): A recurrent GLM implementation.
+    """
+
+    def __init__(
+        self,
+        regularizer: Union[str, Regularizer] = "UnRegularized",
+        regularizer_strength: Optional[float] = None,
+        solver_name: str = None,
+        solver_kwargs: Optional[dict] = None,
+    ):
+        self.regularizer = regularizer
+
+        # check regularizer strength
+        if regularizer_strength is None:
+            warnings.warn(
+                UserWarning(
+                    "Caution: regularizer strength has not been set. Defaulting to 1.0. Please see "
+                    "the documentation for best practices in setting regularization strength."
+                )
+            )
+            self._regularizer_strength = 1.0
+        else:
+            self.regularizer_strength = regularizer_strength
+
+        # no solver name provided, use default
+        if solver_name is None:
+            self.solver_name = self.regularizer.default_solver
+        else:
+            self.solver_name = solver_name
+
+        if solver_kwargs is None:
+            solver_kwargs = dict()
+        self.solver_kwargs = solver_kwargs
+
+    @property
+    def regularizer(self) -> Union[None, Regularizer]:
+        """Getter for the regularizer attribute."""
+        return self._regularizer
+
+    @regularizer.setter
+    def regularizer(self, regularizer: Union[str, Regularizer]):
+        """Setter for the regularizer attribute."""
+        # instantiate regularizer if str
+        if isinstance(regularizer, str):
+            self._regularizer = create_regularizer(name=regularizer)
+        elif isinstance(regularizer, Regularizer):
+            self._regularizer = regularizer
+        else:
+            raise TypeError(
+                f"The regularizer should be either a string from "
+                f"{AVAILABLE_REGULARIZERS} or an instance of `nemos.regularizer.Regularizer`"
+            )
+
+        if isinstance(self._regularizer, UnRegularized):
+            self._regularizer_strength = None
+
+    @property
+    def regularizer_strength(self) -> float:
+        return self._regularizer_strength
+
+    @regularizer_strength.setter
+    def regularizer_strength(self, strength: float):
+        try:
+            # force conversion to float to prevent weird GPU issues
+            strength = float(strength)
```

Comment:
maybe we should tree_map this, for future compatibility (proximal methods allows tree structured regularization, this would enable us to be more flexible with the regularizers.
```suggestion
            strength = jax.tree_util.tree_map(lambda x: jnp.asarray(x, dtype=float), strength) 
```

Reply:
Ignore this and related comments. other PR